chain/events/filter/index.go

```diff
@@ -170,42 +361,86 @@ func (ei *EventIndex) CollectEvents(ctx context.Context, te *TipSetEvents, rever
 				return xerrors.Errorf("tipset key cid: %w", err)
 			}
 
-			res, err := stmtEvent.Exec(
+			// check if this event already exists in the database
+			var entryID sql.NullInt64
+			err = tx.Stmt(ei.stmtEventExists).QueryRow(
 				te.msgTs.Height(),          // height
 				te.msgTs.Key().Bytes(),     // tipset_key
 				tsKeyCid.Bytes(),           // tipset_key_cid
 				addr.Bytes(),               // emitter_addr
 				evIdx,                      // event_index
 				em.Message().Cid().Bytes(), // message_cid
 				msgIdx,                     // message_index
-				revert,                     // reverted
-			)
+			).Scan(&entryID)
 			if err != nil {
-				return xerrors.Errorf("exec insert event: %w", err)
+				return xerrors.Errorf("error checking if event exists: %w", err)
 			}
 
-			lastID, err := res.LastInsertId()
-			if err != nil {
-				return xerrors.Errorf("get last row id: %w", err)
-			}
+			if !entryID.Valid {
+				// event does not exist, lets insert it
+				res, err := tx.Stmt(ei.stmtInsertEvent).Exec(
+					te.msgTs.Height(),          // height
+					te.msgTs.Key().Bytes(),     // tipset_key
+					tsKeyCid.Bytes(),           // tipset_key_cid
+					addr.Bytes(),               // emitter_addr
+					evIdx,                      // event_index
+					em.Message().Cid().Bytes(), // message_cid
+					msgIdx,                     // message_index
+					false,                      // reverted
+				)
+				if err != nil {
+					return xerrors.Errorf("exec insert event: %w", err)
+				}
 
-			for _, entry := range ev.Entries {
-				_, err := stmtEntry.Exec(
-					lastID,                      // event_id
-					isIndexedValue(entry.Flags), // indexed
-					[]byte{entry.Flags},         // flags
-					entry.Key,                   // key
-					entry.Codec,                 // codec
-					entry.Value,                 // value
+				entryID.Int64, err = res.LastInsertId()
+				if err != nil {
+					return xerrors.Errorf("get last row id: %w", err)
+				}
+
+				// insert all the entries for this event
+				for _, entry := range ev.Entries {
+					_, err = tx.Stmt(ei.stmtInsertEntry).Exec(
+						entryID.Int64,               // event_id
+						isIndexedValue(entry.Flags), // indexed
+						[]byte{entry.Flags},         // flags
+						entry.Key,                   // key
+						entry.Codec,                 // codec
+						entry.Value,                 // value
+					)
+					if err != nil {
+						return xerrors.Errorf("exec insert entry: %w", err)
+					}
+				}
+			} else {
+				// event already exists, lets mark it as not reverted
+				res, err := tx.Stmt(ei.stmtRestoreEvent).Exec(
```

Comment:
_Technically_, due to the transaction semantics, if any events are recorded, all should be recorded. So we should be able to check:

1. Are any events recorded for this tipset. If so, unmark them as reverted and return.
2. Otherwise, insert them.

But, we can also keep the current logic unless it becomes a performance issue.